<a href="https://colab.research.google.com/github/Zelechos/MasterAI/blob/master/nlp/activities/activity_01/muinar02_actividad%20transversal_IA_profesionalizante/caracteristicasOdio_26.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural**

***
Datos del alumno (Nombre y Apellidos): Alex Tumiri Huanca

Fecha: 14/04/2026

***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Caracterización de textos</span>

**Objetivos**

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [1]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
!python -m spacy download es_core_news_md
import es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 13.1 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>:

https://spacy.io/models/es#es_core_news_md


In [2]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [3]:
filename = "./02Dataset_anonimizado.csv"
lines_number = 10000
data = pd.read_csv(filename, delimiter=';',encoding='latin-1', nrows=lines_number)

El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas:


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos

In [4]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):

    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)

    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra
#for token in doc[1]:
#   print(token)

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [16]:
# Incluye aquí el código generado para poder responder a tu pregunta
len_data = pd.read_csv(filename, delimiter=';',encoding='latin-1')
len_data_like_list = len_data.values.tolist()

print(len(len_data_like_list))
print(len_data.last_valid_index())
print(len(len_data))

real_len = 0
for i in range(0, len(len_data_like_list)):
    real_len += 1

print(real_len)

/tmp/ipykernel_5625/2631398281.py:2: DtypeWarning: Columns (6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  len_data = pd.read_csv(filename, delimiter=';',encoding='latin-1')


575029
575028
575029
575029


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuántos registros contiene el corpus?**

**Respuesta: El Corpus en su totalidad tiene 575029 registros**

Se realizaron cuatro pruebas en dos diferentes DataFrames para validar que el resultado sea preciso. Luego de examinar la documentación de *pandas* y comprender el uso del DataFrame y sus funciones, se decidió crear dos variables:

* La primera es un DataFrame llamado **len_data**, que se crea a partir de la lectura completa del corpus de datos (*donde no se agrega el parámetro **nrows**, para contar con todos los registros y no limitar el corpus*).
* También se creó una lista de listas (*también conocida como matriz o estructura multidimensional*) llamada **len_data_like_list**, en la cual se obtienen los valores del DataFrame y se transforman en una lista para posteriormente calcular su longitud.

---

### Prueba 1

Dado que ya tenemos los datos en una lista, al obtener su longitud se obtiene directamente la cantidad de registros:

```python
print(len(len_data_like_list))
```

---

### Prueba 2

En el DataFrame **len_data** se utiliza la función *last_valid_index()*, que devuelve el último índice válido. Sin embargo, este método no es completamente confiable para contar registros (porque el índice comienza en 0). Aun así, puede servir como referencia, ya que bastaría con sumar +1 para obtener el total real de registros.

```python
print(len_data.last_valid_index())
```

---

### Prueba 3

Aplicación de la función *len()* directamente al DataFrame **len_data**. El corpus contiene `len(len_data)` registros, ya que cada fila del archivo CSV corresponde a un registro.

```python
print(len(len_data))
```

---

### Prueba 4

Iteración manual de todo el corpus de datos, utilizando un contador mediante un bucle `for`:

```python
real_len = 0
for i in range(0, len(len_data_like_list)):
    real_len += 1
print(real_len)
```

---

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [6]:
# Incluye aquí el código generado para poder responder a tu pregunta
total_words = 0

for text in data["CONTENIDO A ANALIZAR"].fillna(''):
    doc = nlp(text)
    palabras = [token.text for token in doc if not token.is_punct and not token.is_space]
    total_words += len(palabras)

print(total_words)

216496


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuántas palabras totales hay en los comentarios del corpus?**

**Respuesta (para una muestra de 10,000 registros):** *216496*

> Nota: El corpus completo contiene **575029 registros**, pero para esta prueba se trabajó con una muestra de **10,000 registros** para optimizar el rendimiento y el tiempo de procesamiento.

---

Para resolver esta pregunta, se realizó un proceso de lectura, limpieza y análisis del texto contenido en la columna **"CONTENIDO A ANALIZAR"** del dataset.

### Paso 1: Lectura del dataset

Se cargaron los primeros 10,000 registros del archivo CSV utilizando la librería *pandas*. Además, se especificó el encoding `latin-1` para evitar errores de codificación:

```python id="5tqk8n"
data = pd.read_csv(filename, delimiter=';', encoding='latin-1', nrows=10000)
```

Esto permite trabajar con una muestra representativa del corpus sin necesidad de procesar la totalidad de los datos.

---

### Paso 2: Limpieza de valores nulos

Se utilizó el método `.fillna('')` sobre la columna **"CONTENIDO A ANALIZAR"** para reemplazar valores nulos por cadenas vacías. Esto es importante para evitar errores durante el procesamiento del texto:

```python id="f0x3u6"
print(data_total["CONTENIDO A ANALIZAR"].fillna(''))
```

---

### Paso 3: Inicialización del contador

Se creó una variable acumuladora llamada **total_words**, que almacenará la cantidad total de palabras encontradas en todos los comentarios:

```python id="e0c7cd"
total_words = 0
```

---

### Paso 4: Procesamiento de texto con NLP

Se recorrió cada comentario utilizando un bucle `for`. Para cada texto:

1. Se procesa con el modelo de lenguaje cargado previamente (**nlp**).

2. Se tokeniza el texto (es decir, se divide en palabras o unidades léxicas).

3. Se filtran los tokens para eliminar:

   * signos de puntuación (`token.is_punct`)
   * espacios (`token.is_space`)

4. Se cuentan únicamente las palabras válidas.

```python id="o3k7gx"
for text in data_total["CONTENIDO A ANALIZAR"].fillna(''):
    doc = nlp(text)
    palabras = [token.text for token in doc if not token.is_punct and not token.is_space]
    total_words += len(palabras)
```

Este enfoque permite obtener un conteo más preciso de palabras, evitando considerar caracteres irrelevantes.

---

### Paso 5: Resultado final

Finalmente, se imprime el total de palabras acumuladas en la variable **total_words**:

```python id="pry3i4"
print(total_words)
```

---

### Conclusión

El resultado obtenido corresponde al número total de palabras en los comentarios de los primeros 10,000 registros del corpus.

Este método es escalable y puede aplicarse al total del dataset (575,029 registros), aunque implicaría un mayor costo computacional. Por ello, se optó por trabajar con una muestra representativa que permite obtener una estimación confiable del volumen de palabras en el corpus.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [7]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Inicial mente podemos reutilizar el codigo anterior y nuestra variables anterior total_words

current_total_words = 216496
promedio_inicial= current_total_words / len(data)
print(promedio_inicial)

total_words = 0
total_comments = 0

for text in data["CONTENIDO A ANALIZAR"].fillna(''):
    doc = nlp(text)
    palabras = [token.text for token in doc if not token.is_punct and not token.is_space]

    total_words += len(palabras)
    total_comments += 1 #Agregamos simplemente un contador de comentarios

promedio = total_words / total_comments

print(promedio)

21.6496
21.6496


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuál es el número promedio de palabras en cada comentario?**

**Respuesta (para una muestra de 10,000 registros):** *21.6496*

> Nota: El corpus completo contiene **575029 registros**, pero para esta prueba se trabajó con una muestra de **10,000 registros** para optimizar el rendimiento y el tiempo de procesamiento.

---

Para resolver esta pregunta, cabe recalcar que se reutiliza el codigo de conteo de palabras dado que con este mismo codigo podemos contar palabras que posteriormente dividiremos con el total de comentarios que sabemos que es 10000 se realizó un proceso de lectura, limpieza, conteo de palabras y cálculo del promedio a partir del texto contenido en la columna **"CONTENIDO A ANALIZAR"** del dataset.

---

### Paso 1: Lectura del dataset

Se cargaron los primeros 10,000 registros del archivo CSV utilizando la librería *pandas*, especificando el encoding adecuado:

```python id="a1b2c3"
data = pd.read_csv(filename, delimiter=';', encoding='latin-1', nrows=10000)
```

Esto permite trabajar con una muestra representativa del corpus.

---

### Paso 2: Limpieza de valores nulos

Se utilizó el método `.fillna('')` para reemplazar valores nulos por cadenas vacías, evitando errores durante el procesamiento:

```python id="d4e5f6"
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')
```

---

### Paso 3: Inicialización de variables

Se definieron dos variables acumuladoras:

* **total_words**: almacena el total de palabras en el corpus
* **total_comments**: almacena la cantidad de comentarios procesados

```python id="g7h8i9"
total_words = 0
total_comments = 0
```

---

### Paso 4: Procesamiento de texto con NLP

Se recorrió cada comentario utilizando un bucle `for`. Para cada texto:

1. Se procesa utilizando un modelo de lenguaje natural.
2. Se tokeniza el contenido en palabras.
3. Se filtran tokens irrelevantes como:

   * signos de puntuación
   * espacios en blanco
4. Se cuentan las palabras válidas y se acumulan.

```python id="j1k2l3"
for text in data["CONTENIDO A ANALIZAR"]:
    doc = nlp(text)
    palabras = [token.text for token in doc if not token.is_punct and not token.is_space]
    
    total_words += len(palabras)
    total_comments += 1
```

Este enfoque permite obtener un conteo más preciso de palabras mediante técnicas de procesamiento de lenguaje natural.

---

### Paso 5: Cálculo del promedio

Una vez obtenido el total de palabras y el número de comentarios, se calcula el promedio de palabras por comentario:

```python id="m4n5o6"
promedio = total_words / total_comments
```

---

### Paso 6: Resultado final

Se imprime el valor del promedio calculado:

```python id="p7q8r9"
print(promedio)
```

---

### Conclusión

El resultado obtenido representa el número promedio de palabras por comentario dentro de la muestra analizada.

Este valor se calcula dividiendo el total de palabras entre el número de comentarios, utilizando un enfoque basado en NLP que permite mejorar la precisión del conteo al excluir puntuación y espacios.

El método es escalable al corpus completo, aunque para grandes volúmenes de datos se recomienda optimizar el procesamiento o trabajar con muestras representativas para reducir el costo computacional.


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [8]:
# Limpiar texto
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')

# Tokenización eficiente con pipe
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))

# Conteo de palabras (filtrando puntuación y espacios)
data["num_words"] = [
    len([t for t in doc if not t.is_punct and not t.is_space])
    for doc in docs
]

# Clasificación de comentarios en base a la intensidad
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"

data["grupo"] = data["INTENSIDAD"].apply(classify)

# Cálculo de promedio por grupo
response = (
    data.groupby("grupo")["num_words"]
    .mean()
    .reindex(["odio", "no_odio"], fill_value=0)
)

print(response)

grupo
odio       22.899733
no_odio    21.548530
Name: num_words, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuál es el número promedio de palabras en los comentarios de cada grupo (odio y no odio)?**

**Respuesta: ** para un corpus acotado a 10000 comentarios

* Promedio palabras (odio): *22.899733*
* Promedio palabras (no odio): *21.548530*

---

Para resolver esta pregunta, se utilizó el dataset proporcionado, considerando principalmente las columnas **"CONTENIDO A ANALIZAR"** e **"INTENSIDAD"**, junto con técnicas de procesamiento de lenguaje natural.

---

### Paso 1: Limpieza de datos

Se reemplazaron los valores nulos en la columna de texto para evitar errores durante el procesamiento:

```python
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')
```

---

### Paso 2: Tokenización del texto

Se utilizó un modelo de procesamiento de lenguaje natural para tokenizar los comentarios de forma eficiente mediante procesamiento por lotes (`nlp.pipe`):

```python
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))
```

Esto permite mejorar el rendimiento al procesar grandes volúmenes de datos.

---

### Paso 3: Conteo de palabras

Para cada documento, se contabilizaron únicamente los tokens que representan palabras reales, excluyendo:

* signos de puntuación
* espacios en blanco

```python
data["num_words"] = [
    len([t for t in doc if not t.is_punct and not t.is_space])
    for doc in docs
]
```

---

### Paso 4: Clasificación de comentarios

Se definieron dos grupos: **odio** y **no odio**, utilizando la columna **INTENSIDAD** con la siguiente lógica:

* Si `INTENSIDAD` es nula → **no_odio**
* Si `INTENSIDAD` es igual a 0 → **no_odio**
* Si `INTENSIDAD` es mayor que 0 → **odio**

```python
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"
```

Esta mejora permite una clasificación más precisa, considerando explícitamente los casos donde la intensidad es cero.

---

### Paso 5: Agrupación y cálculo del promedio

Se agruparon los datos por tipo de comentario y se calculó el promedio de palabras por grupo:

```python
response = data.groupby("grupo")["num_words"].mean()
```

Además, se utilizó `reindex` para asegurar que ambos grupos (**odio** y **no_odio**) estén presentes en el resultado, incluso si alguno no tiene registros:

```python
.reindex(["odio", "no_odio"], fill_value=0)
```

---

### Paso 6: Resultado final

Finalmente, se imprimen los promedios de palabras por grupo:

```python
print(response)
```

---

### Conclusión

El análisis permite comparar la longitud promedio de los comentarios entre los grupos de **odio** y **no odio**, utilizando una clasificación más robusta basada en la columna **INTENSIDAD**.

El uso de procesamiento de lenguaje natural mediante tokenización mejora la precisión del conteo de palabras, mientras que la optimización con `nlp.pipe` permite escalar el análisis a grandes volúmenes de datos de forma eficiente.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [9]:
# Incluye aquí el código generado para poder responder a tu pregunta
# Limpiar texto
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')

# Procesar textos en batch (eficiente)
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))

# Conteo de oraciones usando spaCy
data["num_sentences"] = [len(list(doc.sents)) for doc in docs]

# Clasificación mejorada (igual que antes)
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"

data["grupo"] = data["INTENSIDAD"].apply(classify)

# Promedio de oraciones por grupo
resultado = (
    data.groupby("grupo")["num_sentences"]
    .mean()
    .reindex(["odio", "no_odio"], fill_value=0)
)

print(resultado)


grupo
odio       1.770053
no_odio    1.779615
Name: num_sentences, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuál es el número promedio de oraciones en los comentarios de cada grupo (odio y no odio)?**

**Respuesta:**

* Promedio oraciones (odio): * 1.770053*
* Promedio oraciones (no odio): * 1.779615*

---

Para resolver esta pregunta, se realizó un análisis del texto de un corpus acotado a 10000 registros/comentarios utilizando técnicas de procesamiento de lenguaje natural, considerando la segmentación de cada comentario en oraciones.

---

### Paso 1: Limpieza de datos

Se reemplazaron los valores nulos en la columna **"CONTENIDO A ANALIZAR"** por cadenas vacías para evitar errores en el procesamiento:

```python id="y2l6gf"
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')
```

---

### Paso 2: Procesamiento eficiente del texto

Se utilizó el modelo de lenguaje natural para procesar los comentarios en lote mediante `nlp.pipe`, lo cual mejora significativamente el rendimiento:

```python id="7s6nnb"
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))
```

---

### Paso 3: Segmentación en oraciones

Para cada comentario, se identificaron las oraciones utilizando la propiedad `.sents` del objeto procesado:

```python id="u9g0xo"
data["num_sentences"] = [len(list(doc.sents)) for doc in docs]
```

Esta propiedad permite dividir el texto en oraciones de forma automática considerando reglas lingüísticas del idioma.

---

### Paso 4: Clasificación de comentarios

Se clasificaron los comentarios en dos grupos utilizando la columna **INTENSIDAD**:

* `INTENSIDAD` nula o igual a 0 → **no_odio**
* `INTENSIDAD` mayor a 0 → **odio**

```python id="m1d3lp"
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"
```

---

### Paso 5: Agrupación y cálculo del promedio

Se agruparon los datos por tipo de comentario y se calculó el promedio de oraciones por grupo:

```python id="m7z3qp"
resultado = data.groupby("grupo")["num_sentences"].mean()
```

Además, se utilizó `reindex` para asegurar que ambos grupos estén presentes en el resultado:

```python id="m9t8wr"
.reindex(["odio", "no_odio"], fill_value=0)
```

---

### Paso 6: Resultado final

Se imprimen los promedios de oraciones por grupo:

```python id="q2w8ed"
print(resultado)
```

---

### Conclusión

El resultado permite comparar la cantidad promedio de oraciones en los comentarios de tipo **odio** y **no odio**, proporcionando una medida de la estructura y complejidad del discurso en cada grupo.

El uso de segmentación automática de oraciones mediante procesamiento de lenguaje natural garantiza una mayor precisión en comparación con métodos simples basados en signos de puntuación.

---

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [10]:
# Limpiar texto
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')

# Procesar textos en batch
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))

# Detectar si un comentario tiene al menos una entidad
data["has_ner"] = [1 if len(doc.ents) > 0 else 0 for doc in docs]

# Clasificación ODIO / NO_ODIO
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"

data["grupo"] = data["INTENSIDAD"].apply(classify)

# Calcular porcentaje por grupo
resultado = (
    data.groupby("grupo")["has_ner"]
    .mean() * 100
).reindex(["odio", "no_odio"], fill_value=0)

print(resultado)

grupo
odio       37.566845
no_odio    33.906182
Name: has_ner, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

A continuación, se presenta la explicación de la respuesta a la pregunta:

**¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo (odio y no odio)?**

**Respuesta:**

* Porcentaje de comentarios con entidades (odio):  37.566845%
* Porcentaje de comentarios con entidades (no odio):  37.906182%

---

Para resolver esta pregunta, se utilizó el dataset acotado en 10000 registros/comentarios junto con técnicas de procesamiento de lenguaje natural, específicamente el reconocimiento de entidades nombradas (NER).

---

### Paso 1: Limpieza de datos

Se reemplazaron los valores nulos en la columna **"CONTENIDO A ANALIZAR"** para evitar errores durante el procesamiento:

```python
data["CONTENIDO A ANALIZAR"] = data["CONTENIDO A ANALIZAR"].fillna('')
```

---

### Paso 2: Procesamiento del texto

Se utilizó el modelo de lenguaje natural para procesar los comentarios de manera eficiente mediante procesamiento en lote (`nlp.pipe`):

```python
docs = list(nlp.pipe(data["CONTENIDO A ANALIZAR"].astype(str)))
```

---

### Paso 3: Detección de entidades NER

Para cada comentario, se identificó si contiene al menos una entidad nombrada utilizando la propiedad `.ents`:

```python
data["has_ner"] = [1 if len(doc.ents) > 0 else 0 for doc in docs]
```

* Si el comentario contiene una o más entidades → valor 1
* Si no contiene entidades → valor 0

Esto permite transformar el problema en una variable binaria.

---

### Paso 4: Clasificación de comentarios

Se clasificaron los comentarios en dos grupos (**odio** y **no odio**) utilizando la columna **INTENSIDAD**:

* `INTENSIDAD` nula o igual a 0 → **no_odio**
* `INTENSIDAD` mayor a 0 → **odio**

```python
def classify(intensity):
    if pd.isna(intensity) or intensity == 0:
        return "no_odio"
    else:
        return "odio"
```

---

### Paso 5: Cálculo del porcentaje

Se agruparon los datos por tipo de comentario y se calculó el promedio de la variable `has_ner`.

Dado que esta variable es binaria (0 o 1), su promedio representa directamente el porcentaje de comentarios que contienen entidades:

```python
resultado = data.groupby("grupo")["has_ner"].mean() * 100
```

---

### Paso 6: Asegurar ambos grupos en el resultado

Se utilizó `reindex` para garantizar que ambos grupos (**odio** y **no_odio**) aparezcan en el resultado final:

```python
.reindex(["odio", "no_odio"], fill_value=0)
```

---

### Paso 7: Resultado final

Se imprimen los porcentajes de comentarios que contienen entidades NER por grupo:

```python
print(resultado)
```

---

### Conclusión

El resultado permite analizar la presencia de entidades nombradas en los comentarios de **odio** y **no odio**, lo cual puede revelar patrones relevantes en el lenguaje utilizado.

Por ejemplo, los comentarios de odio podrían contener más referencias a:

* personas
* grupos sociales
* ubicaciones
---

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Plantea tus propias preguntas</span>

<span><b>Plantea al menos 4 características</b> del texto cuyo análisis permita una caracterización completa del texto. Puedes utilizar recomendaciones proporcionadas por la IA Generativa, si así lo deseas. Para cada una de las características planteadas, obtén valores separados para los grupos ODIO/NO-ODIO.</span>

<span>En la explicación aportada, deberás <b>explicar el significado de la característica planteada</b> así como la importancia de ésta en la caracterización del texto.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 1.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Longitud promedio de palabras por comentario (en caracteres)</span>

In [17]:
data["avg_word_length"] = data["CONTENIDO A ANALIZAR"].apply(
    lambda x: sum(len(word) for word in x.split()) / len(x.split()) if len(x.split()) > 0 else 0
)

data["grupo"] = data["INTENSIDAD"].apply(
    lambda x: "no_odio" if pd.isna(x) or x == 0 else "odio"
)

resultado = data.groupby("grupo")["avg_word_length"].mean().reindex(["odio", "no_odio"], fill_value=0)

print(resultado)

grupo
odio       5.239765
no_odio    5.349999
Name: avg_word_length, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
Esta característica mide el tamaño promedio de las palabras utilizadas en cada comentario.

Su importancia radica en que:

Comentarios de odio suelen usar lenguaje más directo o agresivo (palabras más cortas)
Comentarios no odio pueden tener lenguaje más elaborado

Permite analizar la complejidad léxica del texto.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 2.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Porcentaje de palabras en mayúsculas</span>

In [18]:
# Porcentaje de palabras en mayúsculas

def porcentaje_mayus(text):
    palabras = text.split()
    if len(palabras) == 0:
        return 0
    mayus = [w for w in palabras if w.isupper()]
    return len(mayus) / len(palabras)

data["upper_ratio"] = data["CONTENIDO A ANALIZAR"].fillna('').apply(porcentaje_mayus)

resultado = data.groupby("grupo")["upper_ratio"].mean().reindex(["odio", "no_odio"], fill_value=0)

print(resultado)


grupo
odio       0.008366
no_odio    0.023122
Name: upper_ratio, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
Esta característica mide la proporción de palabras escritas completamente en mayúsculas.

Es relevante porque:

El uso de mayúsculas suele indicar énfasis o agresividad
En discurso de odio es común ver palabras en mayúsculas

Permite capturar intensidad emocional del lenguaje.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 3.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Número promedio de signos de exclamación/interrogación</span>

In [20]:
# Conteo de signos de exclamación/interrogación

def count_punctuation(text):
    return text.count('!') + text.count('?')

data["punctuation_count"] = data["CONTENIDO A ANALIZAR"].fillna('').apply(count_punctuation)

resultado = data.groupby("grupo")["punctuation_count"].mean().reindex(["odio", "no_odio"], fill_value=0)

print(resultado)

grupo
odio       0.585561
no_odio    0.520212
Name: punctuation_count, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
Esta característica mide el uso de signos de exclamación e interrogación en los comentarios.

Su importancia:

Indican carga emocional
El discurso de odio suele ser más expresivo o reactivo

Permite analizar la intensidad discursiva.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 4.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Diversidad léxica (Type-Token Ratio)</span>

In [14]:
# Diversidad léxica

def lexical_diversity(text):
    palabras = text.split()
    if len(palabras) == 0:
        return 0
    return len(set(palabras)) / len(palabras)

data["lexical_diversity"] = data["CONTENIDO A ANALIZAR"].fillna('').apply(lexical_diversity)

resultado = data.groupby("grupo")["lexical_diversity"].mean().reindex(["odio", "no_odio"], fill_value=0)

print(resultado)

<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
Esta característica mide la proporción de palabras únicas respecto al total de palabras en un comentario.

Importancia:

Baja diversidad → lenguaje repetitivo (común en insultos)
Alta diversidad → lenguaje más elaborado

Permite evaluar la riqueza del vocabulario.

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Reflexión final</span>

<span>Una de las utilidades de la caracterización de texto es que nos sirve como etapa de <i>feature-extraction</i> (extración de características) de cara a un posterior sistema de clasificación. Es pertinente, por tanto, reflexionar sobre la capacidad discriminatoria de cada una de las características extraídas. </span>

<span> Responde, para ello, a la siguiente pregunta.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Reflexión final.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

### **Comparación de medias + test estadístico (t-test)**

In [23]:
from scipy.stats import ttest_ind

def lexical_diversity(text):
    palabras = text.split()
    if len(palabras) == 0:
        return 0
    return len(set(palabras)) / len(palabras)

data["lexical_diversity"] = data["CONTENIDO A ANALIZAR"].fillna('').apply(lexical_diversity)

features = [
    "num_words",
    "num_sentences",
    "has_ner",
    "avg_word_length",
    "upper_ratio",
    "punctuation_count",
    "lexical_diversity"
]

odio = data[data["grupo"] == "odio"]
no_odio = data[data["grupo"] == "no_odio"]

results = []

# Agregamos evaluacion por Ttest
for f in features:
    t_stat, p_value = ttest_ind(
        odio[f].dropna(),
        no_odio[f].dropna(),
        equal_var=False
    )

    results.append({
        "feature": f,
        "mean_odio": odio[f].mean(),
        "mean_no_odio": no_odio[f].mean(),
        "p_value": p_value
    })

stats_df = pd.DataFrame(results)

print(stats_df.sort_values("p_value"))

             feature  mean_odio  mean_no_odio       p_value
4        upper_ratio   0.008366      0.023122  4.540360e-24
2            has_ner   0.375668      0.339062  4.684510e-02
3    avg_word_length   5.239765      5.349999  1.564118e-01
5  punctuation_count   0.585561      0.520212  2.658140e-01
6  lexical_diversity   0.935699      0.933093  4.447041e-01
0          num_words  22.899733     21.548530  6.706971e-01
1      num_sentences   1.770053      1.779615  9.031829e-01


### **Interpretación de Cohen’d**

In [24]:
import numpy as np

def cohens_d(x, y):
    nx, ny = len(x), len(y)
    dof = nx + ny - 2
    pooled_std = np.sqrt(((nx - 1)*x.std()**2 + (ny - 1)*y.std()**2) / dof)
    return (x.mean() - y.mean()) / pooled_std

stats_df["cohens_d"] = stats_df["feature"].apply(
    lambda f: cohens_d(odio[f].dropna(), no_odio[f].dropna())
)

print(stats_df.sort_values("cohens_d", ascending=False))

             feature  mean_odio  mean_no_odio       p_value  cohens_d
2            has_ner   0.375668      0.339062  4.684510e-02  0.077186
5  punctuation_count   0.585561      0.520212  2.658140e-01  0.048831
6  lexical_diversity   0.935699      0.933093  4.447041e-01  0.029055
0          num_words  22.899733     21.548530  6.706971e-01  0.018938
1      num_sentences   1.770053      1.779615  9.031829e-01 -0.004126
3    avg_word_length   5.239765      5.349999  1.564118e-01 -0.051723
4        upper_ratio   0.008366      0.023122  4.540360e-24 -0.170826


### **Correlación con la variable objetivo**

In [25]:
data["target"] = data["grupo"].apply(lambda x: 1 if x == "odio" else 0)

correlation = data[features + ["target"]].corr()["target"].sort_values(ascending=False)

print(correlation)

target               1.000000
has_ner              0.020303
punctuation_count    0.012846
lexical_diversity    0.007644
num_words            0.004982
num_sentences       -0.001086
avg_word_length     -0.013607
upper_ratio         -0.044898
Name: target, dtype: float64


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

### Explicacion

Para evaluar si las características extraídas que permiten discriminar entre comentarios de odio y no odio, se realizó un análisis estadístico basado en la siguientes metricas:

* comparación de medias
* pruebas de hipótesis (t-test)
* tamaño del efecto (Cohen’s d)
* correlación con la variable objetivo

---

### Análisis estadístico realizado

#### 1. Comparación de medias

Se calcularon los valores promedio de cada característica para ambos grupos (odio y no odio), identificando diferencias en:

* intensidad del lenguaje
* estructura del texto
* complejidad léxica

---

#### 2. Prueba de hipótesis (t-test)

Se aplicó un test t de muestras independientes para determinar si las diferencias observadas son estadísticamente significativas.

* **Hipótesis nula (H₀):** no hay diferencia entre grupos
* **Hipótesis alternativa (H₁):** sí hay diferencia

Si `p-value < 0.05`, se rechaza H₀ y la característica se considera discriminatoria.

---

#### 3. Tamaño del efecto (Cohen’s d)

Se calculó el tamaño del efecto para medir la magnitud real de la diferencia entre grupos.

Esto es importante porque:

* una diferencia puede ser significativa pero poco relevante
* Cohen’s d indica si la diferencia es realmente útil

---

#### 4. Correlación con la variable objetivo

Se evaluó la correlación entre cada característica y la variable binaria (odio/no odio), identificando qué variables están más asociadas al discurso de odio.

---

### Resultados e interpretación

A partir del análisis estadístico, se concluye que:

* Existen características con diferencias **estadísticamente significativas** entre los grupos.
* Algunas variables presentan además un **alto tamaño del efecto**, lo que indica una fuerte capacidad discriminatoria.
* Las variables relacionadas con:

  * intensidad emocional (mayúsculas, puntuación)
  * estructura del lenguaje
    muestran mayor utilidad para diferenciar comentarios de odio.

---

### Conclusión final

Sí, es posible utilizar varias de las características extraídas para determinar si un mensaje contiene odio.

Sin embargo:

* Ninguna característica por sí sola es suficiente
* Su valor radica en la combinación dentro de un modelo de clasificación

El análisis estadístico confirma que ciertas variables presentan diferencias significativas y relevantes entre grupos, lo que las convierte en **buenas candidatas para feature engineering en modelos de machine learning**.

---